cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/nosql/NoSqlStructuredTable.java

```diff
@@ -126,6 +132,87 @@ public CloseableIterator<StructuredRow> scan(Range keyRange, int limit) throws I
                              limit);
   }
 
+
+  /*
+   *  Sorting in memory for the no-sql implementation. We sort post table scan.
+   * */
+  @Override
+  public CloseableIterator<StructuredRow> scan(Range keyRange, int limit, String orderByField, SortOrder sortOrder)
+    throws InvalidFieldException, IOException {
+    LOG.trace("Table {}: Scan range {} with limit {}, sort field {} and sort order {}", schema.getTableId(),
+              keyRange, limit, orderByField, sortOrder);
+    if (!schema.isIndexColumn(orderByField)) {
+      throw new InvalidFieldException(schema.getTableId(), orderByField, "is not an indexed column");
+    }
+    Comparator<StructuredRow> comparator = sortOrder.equals(SortOrder.ASC) ? getComparator(orderByField) :
+      getComparator(orderByField).reversed();
+    PriorityQueue<StructuredRow> rows = new PriorityQueue<>(comparator);
+
+    try (ScannerIterator scannerIterator = new ScannerIterator(getScanner(keyRange), schema)) {
+      while (scannerIterator.hasNext()) {
+        rows.offer(scannerIterator.next());
+      }
+    }
+    // return an iterator for the sorted elements
+    AbstractCloseableIterator<StructuredRow> abstractCloseableIterator = new
+      AbstractCloseableIterator<StructuredRow>() {
+      @Override
+      protected StructuredRow computeNext() {
+        while (!rows.isEmpty()) {
+          return rows.poll();
+        }
+        return endOfData();
+      }
+
+      @Override
+      public void close() {
+        // no-op
+      }
+    };
+    return new LimitIterator(Collections.singleton(abstractCloseableIterator).iterator(), limit);
```

Comment:
nit: would be good to add another constructor to LimitIterator that does this collections wrapping

Reply:
then in that case https://github.com/cdapio/cdap/pull/14554/files#diff-b56e65455ea56dcae705b23d0d2ca23d4be4555e069f2afa27f754568bf64bddR536 is not final right?

Reply:
I'm not sure what the linked diff is pointing to, don't see anything about LimitIterator. I just mean adding a constructor like:

```
  public LimitIterator(CloseableIterator<T> iter, limit) {
    this(Collections.singleton(iter).iterator(), limit);
  }
```

Then you don't need the duplicate code in this class, and it's also more readable.

Reply:
sorry I misunderstood, got it